R/shared-methods.R

```diff
@@ -509,6 +509,158 @@ DataClass <- R6::R6Class(
         source_data_cols = paste(unlist(self$source_data_cols), collapse = ", ")
       )
       return(sum_df)
+    },
+
+    #' @description Run tests on data class. Inherited by child classes so test
+    #' run through each class.
+    #' @param download logical. To download the data (TRUE) or use a snapshot
+    #' (FALSE). Defaults to FALSE.
+    #' @importFrom purrr walk map
+    #' @importFrom dplyr slice_tail
+    #' @importFrom testthat test_that expect_s3_class expect_true expect_error
+    test = function(download = FALSE) {
+      data_name <- paste0(class(self)[1], " at level ", self$level)
+      raw_path <- paste0(
+        "custom_data/", class(self)[1],
+        "_level_", self$level, ".rds"
+      )
+      if (!file.exists(raw_path)) {
+        download <- TRUE
+      }
+      if (download) {
+        test_that(paste0(data_name, " downloads sucessfully"), {
+          self$download()
+          walk(self$data$raw, function(data) {
```

Comment:
there is quite a lot of code here. some of these things might be nicely pulled out as internal functions.